azurerm/resource_arm_sql_firewall_rule.go

```diff
@@ -51,96 +50,86 @@ func resourceArmSqlFirewallRule() *schema.Resource {
-		StartIPAddress:    azure.String(d.Get("start_ip_address").(string)),
-		EndIPAddress:      azure.String(d.Get("end_ip_address").(string)),
-	}
+func resourceArmSqlFirewallRuleCreateOrUpdate(d *schema.ResourceData, meta interface{}) error {
+	sqlFirewallRulesClient := meta.(*ArmClient).sqlFirewallRulesClient
 
-	createResponse, err := createRequest.Execute()
-	if err != nil {
-		return fmt.Errorf("Error creating SQL Server Firewall Rule: %s", err)
-	}
-	if !createResponse.IsSuccessful() {
-		return fmt.Errorf("Error creating SQL Server Firewall Rule: %s", createResponse.Error)
-	}
+	name := d.Get("name").(string)
+	resGroup := d.Get("resource_group_name").(string)
+	serverName := d.Get("server_name").(string)
+	startIpAddress := d.Get("start_ip_address").(string)
+	endIpAddress := d.Get("end_ip_address").(string)
 
-	readRequest := rivieraClient.NewRequest()
-	readRequest.Command = &sql.GetFirewallRule{
-		Name:              d.Get("name").(string),
-		ResourceGroupName: d.Get("resource_group_name").(string),
-		ServerName:        d.Get("server_name").(string),
+	props := sql.FirewallRuleProperties{
+		StartIPAddress: &startIpAddress,
+		EndIPAddress:   &endIpAddress,
+	}
+	parameters := sql.FirewallRule{
+		Name: &name,
+		FirewallRuleProperties: &props,
 	}
 
-	readResponse, err := readRequest.Execute()
+	result, err := sqlFirewallRulesClient.CreateOrUpdate(resGroup, serverName, name, parameters)
 	if err != nil {
-		return fmt.Errorf("Error reading SQL Server Firewall Rule: %s", err)
+		return err
 	}
-	if !readResponse.IsSuccessful() {
-		return fmt.Errorf("Error reading SQL Server Firewall Rule: %s", readResponse.Error)
+
+	if result.ID == nil {
+		return fmt.Errorf("Cannot create sql firewall rule %s (resource group %s) ID", name, resGroup)
 	}
 
-	resp := readResponse.Parsed.(*sql.GetFirewallRuleResponse)
-	d.SetId(*resp.ID)
+	d.SetId(*result.ID)
 
 	return resourceArmSqlFirewallRuleRead(d, meta)
 }
 
 func resourceArmSqlFirewallRuleRead(d *schema.ResourceData, meta interface{}) error {
+	sqlFirewallRulesClient := meta.(*ArmClient).sqlFirewallRulesClient
+
 	id, err := parseAzureResourceID(d.Id())
 	if err != nil {
 		return err
 	}
-	resGroup := id.ResourceGroup
 
-	client := meta.(*ArmClient)
-	rivieraClient := client.rivieraClient
-
-	readRequest := rivieraClient.NewRequestForURI(d.Id())
-	readRequest.Command = &sql.GetFirewallRule{}
+	resGroup := id.ResourceGroup
+	name := id.Path["firewallRules"]
+	serverName := id.Path["servers"]
 
-	readResponse, err := readRequest.Execute()
+	result, err := sqlFirewallRulesClient.Get(resGroup, serverName, name)
 	if err != nil {
-		return fmt.Errorf("Error reading SQL Server Firewall Rule: %s", err)
+		return fmt.Errorf("Error reading SQL firewall rule %s: %v", name, err)
 	}
-	if !readResponse.IsSuccessful() {
-		log.Printf("[INFO] Error reading SQL Server Firewall Rule %q - removing from state", d.Id())
+	if result.Response.StatusCode == http.StatusNotFound {
```

Comment:
can we move this check inside the `if err != nil` statement? in addition it's better to use [this wrapper](https://github.com/terraform-providers/terraform-provider-azurerm/blob/2c73509a1b276a4439c9da5d722ffd59bb011a9e/azurerm/resource_arm_sql_server.go#L123) because if response.Response is nil, this'll currently crash